src/toxcore/dht_new/client.rs

```diff
@@ -0,0 +1,236 @@
+/*
+    Copyright © 2017 Zetok Zalbavar <zexavexxe@gmail.com>
+    Copyright © 2018 Namsoo CHO <nscho66@gmail.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+/*!
+Hold peer info. connected.
+The object of this struct is an object per a peer.
+*/
+
+use futures::{Sink, Future};
+use futures::sync::mpsc;
+use tokio_io::IoFuture;
+
+//use std::collections::VecDeque;
+use std::io::{ErrorKind, Error};
+use std::net::SocketAddr;
+
+use toxcore::crypto_core::*;
+use toxcore::dht_new::packet::*;
+use toxcore::dht_new::codec::*;
+
+/// Shorthand for the transmit half of the message channel.
+type Tx = mpsc::UnboundedSender<DhtUdpPacket>;
+
+/// peer info.
+#[derive(Clone, Debug)]
+pub struct Client {
+    /// Public key of peer
+    pub pk: PublicKey,
+    /// socket address of peer
+    pub addr: SocketAddr,
+    /// last sent ping_id to check PingResponse is correct
+    pub ping_id: u64,
+    // /// last sent NodesRequest id to check NodesResponse is correct
+    // pub nodes_id: u64,
+    // /// last sent NatPingRequest id to check NatPingResponse is correct
+    // pub nat_ping_id: u64,
+    /// precomputed key for this peer
+    pub precomputed_key: PrecomputedKey,
+}
+
+impl Client {
+    /// create Client object
+    pub fn new(precomputed_key: PrecomputedKey, pk: PublicKey, addr: SocketAddr) -> Client {
+        Client {
+            pk: pk,
+            addr: addr,
+            precomputed_key: precomputed_key,
+            ping_id: 0xFFFFFFFFFFFFFFFF,
```

Comment:
why oxfff?

Reply:
Because, can't set ping_id of Client to desired value.
So, can't check in test code.

Reply:
But why did you chose such a value?

Reply:
As you can see at https://github.com/NamsooCho/tox-1/blob/7293bfb67bfd02087f4463a438c2a64fcf01df1a/src/toxcore/dht_new/server.rs#L217-L223

0 is checked

Reply:
https://github.com/NamsooCho/tox-1/blob/7293bfb67bfd02087f4463a438c2a64fcf01df1a/src/toxcore/dht_new/server.rs#L466-L468

Reply:
Client is moved to inside of handle_packet().
So, can't set ping_id.

Reply:
https://github.com/NamsooCho/tox-1/blob/7293bfb67bfd02087f4463a438c2a64fcf01df1a/src/toxcore/dht_new/server.rs#L134-L138

Reply:
In server you check that it should not be 0. But you randomize the id before you actually send ping request: https://github.com/tox-rs/tox/pull/25/files#diff-5c3240e241a145d5d67ead09105441e7R113  . So is there any reason why you had to chose 0xffffff as initial value?

Reply:
https://github.com/NamsooCho/tox-1/blob/7293bfb67bfd02087f4463a438c2a64fcf01df1a/src/toxcore/dht_new/server.rs#L442

https://github.com/NamsooCho/tox-1/blob/7293bfb67bfd02087f4463a438c2a64fcf01df1a/src/toxcore/dht_new/server.rs#L461

In previous commit,
"client.ping_id = prs.id" was possible.

Now, client is not exposed to external.

Reply:
So I don't know how to write test code for handle_packet(), especially in *Response()

Reply:
But you can't just create a new client for every packet here: https://github.com/NamsooCho/tox-1/blob/7293bfb67bfd02087f4463a438c2a64fcf01df1a/src/toxcore/dht_new/server.rs#L137
You have to store somehow client or it's ping_id. The server should remember this in order to be able to handle ping responses correctly. So if server store it you can set it in tests.

Reply:
Ah !!!
The Client object should be created at request time.
When receiving request from other peer, just response with same ping_id.
Ping_id is needed when response is checked.
Client object is enough to be kept just one transaction of req/res.
Right?

Reply:
> Ping_id is needed when response is checked.
Yes.

> Client object is enough to be kept just one transaction of req/res.
Server sends pings to each node from it's kbucket periodically. So it should store somehow ping ids for each node from kbucket. And when server receives ping response it checks if ping id is the same as stored or no. If yes then response is considered as correct. Perhaps we shouldn't store client (do we need this entity at all?) but only ping ids for each node in kbucket.

Reply:
From implementing perspective, Client is useful for containing Tx(mpsc), SocketAddr, pk, ping_id,...
Current kbucket contains only SocektAddr, PublicKey.
Maybe It is better to extend kbucket.

Reply:
> From implementing perspective, Client is useful for containing Tx(mpsc), SocketAddr, pk, ping_id,...


And why does it contain ping_id? It's useless if you have new client for every packet.

Reply:
Got it!!
Thanks for your comments.
I will try to remove Client struct.

Reply:
And please, don't init ping_id with 0xffff